source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,659 @@
+unittest
+{
+  pure @nogc int answer(lazy string question);
+  alias F = answer; // typically F is a template argument
+  static assert(
+    refract!(F, "F").mixture ==
+    "pure @nogc @system std.traits.ReturnType!(F) answer(lazy std.traits.Parameters!(F)[0] _0);");
+}
+
+///
+unittest
+{
+  import std.format;
+  import std.traits;
+
+  interface GrandTour
+  {
+    pure int foo() immutable;
+    @nogc @trusted nothrow ref int foo(out real, return ref int, lazy int) const;
+    @safe shared scope void bar(scope Object);
+  }
+
+  class Mock(Interface) : Interface
+  {
+    static foreach (member; __traits(allMembers, Interface)) {
+        static foreach (fun; __traits(getOverloads, Interface, member)) {
+          mixin({
+              enum Model = refract!(fun, "fun");
+              if (is(ReturnType!fun == void)) {
+                return Model.setBody("{}").mixture;
+              } else if (Model.attributes & FunctionAttribute.ref_) {
+                return Model.setBody(q{{
+                    static %s rv;
+                    return rv;
+                  }}.format(Model.returnType)).mixture;
+              } else {
+                return Model.setBody(q{{
+                    return %s.init;
+                  }}.format(Model.returnType)).mixture;
+              }
+            }());
+        }
+    }
+  }
+
+  GrandTour mock = new Mock!GrandTour;
+  real x;
+  int i, l;
+  mock.foo(x, i, l++) = 1;
+  assert(mock.foo(x, i, l++) == 1);
+  assert(l == 0);
+}
+
+private enum isAggregate(T...) =
+  is(T[0] == struct) || is(T[0] == union) || is(T[0] == class)
+  || is(T[0] == interface);
+
+/**
+    A struct capturing all the properties of a function.
+*/
+
+private mixin template replaceAttribute(string Name)
+{
+  alias Struct = typeof(this);
+  mixin(
+    "Struct copy = {",
+    {
+      string[] mixture;
+      foreach (member; __traits(allMembers, Struct)) {
+        if (__traits(getOverloads, Struct, member).length == 0) {
+          //pragma(msg, member);
+          mixture ~= member ~ ":" ~ (member == Name ? "value" : member);
+        }
+      }
+      return mixture.join(",\n");
+    }(),
+    "};"
+  );
+}
+
+
+immutable struct Function
+{
+
+  /**
+     A string that evaluates to a function symbol.
+   */
+
+  string localSymbol;
+
+  /**
+     Function name. Initial value: `__traits(identifier, fun)`.
+   */
+
+  string name;
+
+  /**
+     Return a new Function object with the name attribute set to value.
+  */
+
+  Function setName(string value)
+  {
+    mixin replaceAttribute!"name";
+    return copy;
+  }
+
+  ///
+  unittest
+  {
+    pure @nogc int answer();
+    mixin(refract!(answer, "answer").setName("ultimateAnswer").mixture);
+    static assert(
+      __traits(getAttributes, ultimateAnswer) ==
+      __traits(getAttributes, answer));
+  }
+
+  /**
+     Index of function in a set of overloads. Valid only if Function represents
+     a function (not a function pointer), and was refracted from a module.
+   */
```

Comment:
Should this have a value of -1 if it's not valid? Or is there a way to know if it's valid or not? I'd if you pass a `Function` object to some other function and inspect this `index`, how do you know?

And how about if it's a template function?

And what is the use-case for getting the `index` and/or setting the `index` of a function?

Reply:
> Should this have a value of -1 if it's not valid? Or is there a way to know if it's valid or not? If you pass a `Function` object to some other function and inspect this `index`, how do you know?

Good point. `-1` it is.

> And how about if it's a template function?

Alas functions resulting from function template instantiation are not addressable like this. D lacks uniformity here.

> And what is the use-case for getting the `index` and/or setting the `index` of a function?

the `refract` overloads do that. In fact, the first `refract`, the one that takes a straight alias to a function, is useless in presence of overloads. Just like there is no easy way to form an alias to an overloaded function. 

Consider:

```
int answer() { return 42; }
//string answer() { return "forty two"; }

void main() {
    import std.stdio;
    alias f = answer;
    writeln(f());
}
```

Now uncomment the `void answer`. 

```
ptrovl.d(7): Error: `ptrovl.answer` called with argument types `()` matches both:
ptrovl.d(1):     `ptrovl.answer()`
and:
ptrovl.d(2):     `ptrovl.answer()`
```

The easiest workaround is:

```
int answer() { return 42; }
string answer() { return "forty two"; }

void main() {
    import std.stdio;
    alias f = __traits(getOverloads, mixin(__MODULE__), "answer")[0];
    writeln(f()); // 42
}
``

In fact we could have somewhere in bolts a helper that returns an alias to an overload given a function name and its signature, e.g.:

```
    alias f1 = functionAlias!("answer", int function()); // return 1st overload
    alias f2 = functionAlias!("answer", string function()); // return 2ndoverload
```